torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp

```diff
@@ -0,0 +1,637 @@
+    auto axis_i = i - 1;
+    if (reference1->axis(axis_i)->isBroadcast() ||
+        reference1->axis(axis_i)->isReduction()) {
+      continue;
+    }
+    if (lhs_i == -1) {
+      lhs_i = axis_i;
+    } else {
+      reference1->merge(axis_i, lhs_i);
+      lhs_i = axis_i;
+    }
+  }
+  reference1->split(0, 1);
+  // [merged_dim, 1, unmerged dims..., tile1, tile2]
+
+  // parallelize non-tile dimensions
+  reference1->axis(1)->parallelize(ParallelType::Unswitch);
+  reference1->axis(0)->parallelize(ParallelType::BIDx);
+  // [BIDx, Unswitch, unmerged dims..., tile1, tile2]
+
+  // Propagate transformations so far to the entire DAG
+  TransformPropagator propagator(reference1);
+  MaxRootDomainInfoSpanningTree entire_dag(reference1);
+  entire_dag.traverse(&propagator);
+  scheduler_utils::parallelizeAllLike(reference1);
+
+  // For a transpose scheduling, all we need is to bind threadIdx.x differently
+  // for inputs and outputs. This swap of binding could happen at any tensor on
+  // the path from input to output, especially, it does not have to be in the
+  // transpose tensor. Here, we naively do the binding swap at cached
+  // input/output for simplicity. We might need to find a better set of swap
+  // tensors in the future to reduce shared memory usage.
+
+  // transform tile for vectorization/unroll
+  // See note [vectorization and unroll of input and output]
+  int64_t num_elems_per_tile = params.tile_size1 * params.tile_size2;
+  int64_t num_elems_per_thread = num_elems_per_tile / kThreadsPerBlock;
+  int64_t unroll_factor1 = num_elems_per_thread / params.vectorize_factor1;
+  int64_t unroll_factor2 = num_elems_per_thread / params.vectorize_factor2;
+
+  int warp_size = at::cuda::getCurrentDeviceProperties()->warpSize;
+
+  // schedule group 2
+  int pos = reference2->nDims() - 2;
+  // [..., tile1, tile2]
+  reference2->merge(pos);
+  reference2->split(pos, params.vectorize_factor2);
+  reference2->split(pos, warp_size);
+  reference2->split(pos, unroll_factor2, false);
+  reference2->axis(-1)->parallelize(ParallelType::Vectorize);
+  reference2->axis(-2)->parallelize(ParallelType::TIDx);
+  reference2->axis(-3)->parallelize(ParallelType::TIDy);
+  reference2->axis(-4)->parallelize(ParallelType::Unroll);
+  // [..., Unroll, TIDy, TIDx, Vectorize]
+
+  // Propagate transformations of reference2 to the entire DAG except
+  // group 1. We actually only want to propagate to the fusion outputs, but
+  // inputs and outputs themselves are disconnected, so we have to borrow the
+  // entire DAG and use its spanning tree.
+  {
+    auto all_tvs_except1 = ir_utils::allTvsExcept(
+        fusion,
+        {grouped_inputs_outputs[0].begin(), grouped_inputs_outputs[0].end()});
+    SetSelector selector({all_tvs_except1.begin(), all_tvs_except1.end()});
+    MaxRootDomainInfoSpanningTree entire_dag_except1(reference2, &selector);
+    TransformPropagator propagator(reference2);
+    entire_dag_except1.traverse(&propagator);
```

Comment:
Is group 2 actually fully connected and convex? How does everything in group 2 actually get propagated to? I thought group 1 was the big connected group, but group 2 isn't necessarily.

Reply:
Wouldn't you want to set the entire problem up like group 2, propagate it, parallelize it, etc. Then reschedule based on reference 1 then propagate that information?

Reply:
It has to be a three-step propagation: The first step is to make tile, merge outer dims and parallelize on `referenceX` (`referenceX` can be either `reference1` or `reference2`, that does not make much difference) , and propagate it to the entire graph. The second step is to transform and parallelize the inner dims of `reference2` and propagate to the entire graph except `reference1` (and this propagation set should be fully connected), then the last step is to reschedule based on reference and propagate.

We can not merge step 1 and step 2 into one step, because we want the outer dims to propagate to the entire graph, but we do not want the transformation of inner dims of `reference2` to touch `reference1`.